### ISSUE TITLE
Release the GIL in Cython code

### ISSUE THREAD

**Issue by mrocklin** [pydata/pandas#8882]
Pandas could release the GIL in parts of the Cython code.  This would enable the parallel use of multiple cores with threading.

From off-line conversation with @jreback @cpcloud 

http://lbolla.info/blog/2013/12/23/python-threads-cython-gil

**Comment by TomAugspurger** [comment]
FYI: scinkit-image just did this in https://github.com/scikit-image/scikit-image/pull/1519.

**Comment by mrocklin** [comment]
It happened surprisingly quickly in their case.  I think that it was mostly the result of an afternoon sprint of a couple developers followed by a couple days of cleanup.

**Comment by jreback** [comment]
@cpcloud did some prelim work on this. I don't think its too difficult. Just requires some effort.

**Comment by shoyer** [comment]
Indeed, it mostly just takes a little bit of testing work. Glad to see this being done!

For what it's worth, I looked into releasing the GIL a little bit in the hashtable/index API as part of my IntervalIndex work (https://github.com/pydata/pandas/pull/8707). The main stumbling block I encountered was `pandas.hashtable.Int64Vector` (and the other `Vector` objects). Hopefully someone more clever than me could figure out how to release the GIL on the `append` method by moving the `resize` call to use the C API: https://github.com/pydata/pandas/blob/0aceb38ac1a4659ae298d726bc447b97943163dc/pandas/hashtable.pyx#L68
